clippy_lints/src/loops.rs

```diff
@@ -1140,43 +1140,51 @@ fn detect_same_item_push<'tcx>(
     walk_expr(&mut same_item_push_visitor, body);
     if same_item_push_visitor.should_lint {
         if let Some((vec, pushed_item)) = same_item_push_visitor.vec_push {
-            // Make sure that the push does not involve possibly mutating values
-            if let PatKind::Wild = pat.kind {
-                let vec_str = snippet_with_macro_callsite(cx, vec.span, "");
-                let item_str = snippet_with_macro_callsite(cx, pushed_item.span, "");
-                if let ExprKind::Path(ref qpath) = pushed_item.kind {
-                    if_chain! {
-                        if let Res::Local(hir_id) = qpath_res(cx, qpath, pushed_item.hir_id);
-                        let node = cx.tcx.hir().get(hir_id);
-                        if let Node::Binding(pat) = node;
-                        if let PatKind::Binding(bind_ann, ..) = pat.kind;
-                        if !matches!(bind_ann, BindingAnnotation::RefMut | BindingAnnotation::Mutable);
-                        then {
-                            span_lint_and_help(
-                                cx,
-                                SAME_ITEM_PUSH,
-                                vec.span,
-                                "it looks like the same item is being pushed into this Vec",
-                                None,
-                                &format!(
-                                    "try using vec![{};SIZE] or {}.resize(NEW_SIZE, {})",
-                                    item_str, vec_str, item_str
-                                ),
-                            )
+            let ty = cx.typeck_results().expr_ty(pushed_item);
```

Comment:
We should use the `T` in `Vec<T>` instead of the type of the pushed item, as those could differ if coercion kicks in. For example this would still be linted:
```rust
#[derive(Clone)]
struct S {}

trait T {}
impl T for S {}

fn test() {
    let mut vec: Vec<Box<dyn T>> = Vec::new();
    for _ in 0..10 {
        vec.push(Box::new(S{}));
    }
    vec.resize(20, Box::new(S{})); // => this gives a compilation error, resize not implemented for Vec<Box<dyn T>>
}
``` 

We can get the type parameter using `vec_ty.walk().nth(1).unwrap().expect_ty();`. `vec_ty` can be obtained using `expr_ty` on the `vec` variable.

Reply:
Thanks for your review! I fixed it.